src/svp.1
.\"
.\" CopyLeft (c) 2016-2018 tokiclover <tokiclover@gmail.com>
.\"
.\" Distributed under the terms of the 2-clause BSD License as
.\" stated in the COPYING file that comes with the source files
.\"
.pc
.TH SVP 1 "2019-03-14" "0.15.0" "User Commands Manual"
.SH NAME
svp \-  supervision program run
.SH DESCRIPTION
.B svp
is a utility to run a programm with new software limits and environment.
This utility can be executed as
.B envdir
or
.B envuidgid
or
.B pgrphack
or
.B softlimit
or
.B setlock
or
.B setuidgid
or
.B setsid
with limited options.
See
.B SYNOPSIS
for more information.

.SH SYNOPSIS
.B svp
.RB [\| \-/ | \-\-chroot \|= DIR \|]
.RB [\| \-b | \-\-name \|= ARG \|]
.RB [\| \-c | \-\-core \|= ARG \|]
.RB [\| \-d | \-\-data \|= DIR \|]
.RB [\| \-e | \-\-envdir \|= DIR \|]
.RB [\| \-f | \-\-output \|= ARG \|]
.RB [\| \-G | \-\-egid \|= GROUP \|]
.RB [\| \-U | \-\-euid \|= USER[:GROUP] \|]
.RB [\| \-L | \-\-Lock \|= FILE \|]
.RB [\| \-l | \-\-lock \|= FILE \|]
.RB [\| \-m | \-\-memory \|= ARG \|]
.RB [\| \-n | \-\-nice \|= ARG \|]
.RB [\| \-o | \-\-open \|= ARG \|]
.RB [\| \-p | \-\-proc \|= DIR \|]
.RB [\| \-g | \-\-gig \|= GROUP \|]
.RB [\| \-u | \-\-uid \|= USER[:GROUP] \|]
.RB [\| \-s | \-\-sid \|]
.RB [\| \-P | \-\-pgrp \|]
.RB [\| \-0 | \-\-stdin \|]
.RB [\| \-1 | \-\-stdout \|]
.RB [\| \-2 | \-\-stderr \|]
.RB [\| \-V | \-\-verbose \|]
.BR \| PROG \|
.RB [\| ARGS \|]

.B svp
.RB [\| \-h | \-\-help \|]
.RB [\| \-v | \-\-version \|]

.B envdir
.RB \| DIR \|
.RB \| PROG·\|
.RB [\| ARGS \|]

.B envuidgid
.RB [\| \-U | \-u \ USER[:GROUP] \|]
.RB [\| \-G | \-g \ GROUP \|]
.RB \| PROG·\|
.RB [\| ARGS \|]

.B pgrphack
.RB \| PROG·\|
.RB [\| ARGS \|]

.B softlimit
.RB [\| \-c | \-\-core \ ARG \|]
.RB [\| \-d | \-\-data \ DIR \|]
.RB [\| \-f | \-\-output \ ARG \|]
.RB [\| \-m | \-\-memory \ ARG \|]
.RB [\| \-n | \-\-nice \ ARG \|]
.RB [\| \-o | \-\-open \ ARG \|]
.RB [\| \-p | \-\-proc \ DIR \|]
.RB \| PROG·\|
.RB [\| ARGS \|]

.B setlock
.RB [\| \-L | \-\-Lock \|]
.RB [\| \-l | \-\-lock \|]
.RB \| FILE \|
.RB \| PROG·\|
.RB [\| ARGS \|]

.B setuidgid
.RB [\| \-U | \-u \ USER[:GROUP] \|]
.RB [\| \-G | \-g \ GROUP \|]
.RB \| PROG·\|
.RB [\| ARGS \|]

.B setsid
.RB \| PROG·\|
.RB [\| ARGS \|]

.SH OPTIONS
.TP
.RB [\| \-/ | \-\-chroot \|= DIR \|]
Change root directory to
.I DIR
.TP
.RB [\| \-b | \-\-name \|= ARG \|]
Set the first vector argument, this is the default behavior
.TP
.RB [\| \-c | \-\-core \|= ARG \|]
Set core size to
.I ARG
bytes
.TP
.RB [\| \-d | \-\-data \|= DIR \|]
Set data size to
.I ARG
bytes
.TP
.RB [\| \-e | \-\-envdir \|= DIR \|]
Use
.I DIR
environment directory
.TP
.RB [\| \-f | \-\-output \|= ARG \|]
Set output file size to
.I ARG
bytes
.TP
.RB [\| \-G | \-\-egid \|= GROUP \|]
Set effective group identity
.TP
.RB [\| \-U | \-\-euid \|= USER[:GROUP] \|]
Set effective user, with optional group identity
.TP
.RB [\| \-L | \-\-Lock \|= FILE \|]
Set a lock file without blocking
.TP
.RB [\| \-l | \-\-lock \|= FILE \|]
Set a lock with blocking
.TP
.RB [\| \-m | \-\-memory \|= ARG \|]
Set memory size to
.I ARG
bytes
.TP
.RB [\| \-n | \-\-nice \|= ARG \|]
Set niceness to
.I ARG
increment
.TP
.RB [\| \-o | \-\-open \|= ARG \|]
Set the maximum number of opened file to
.I ARG
which should be equal or above
.B \_POSIX_OPEN_MAX
.TP
.RB [\| \-p | \-\-proc \|= DIR \|]
Set the maximum of user process to
.I ARG
.TP
.RB [\| \-g | \-\-gig \|= GROUP \|]
Set the group identity
.TP
.RB [\| \-u | \-\-uid \|= USER[:GROUP] \|]
Set the user identity, with an optional group identity
.TP
.RB [\| \-s | \-\-sid \|]
Use a new session
.TP
.RB [\| \-P | \-\-pgrp \|]
Use a new programm group
.TP
.RB [\| \-0 | \-\-stdin \|]
Close standard inoput
.TP
.RB [\| \-1 | \-\-stdout \|]
Close standard output
.TP
.RB [\| \-2 | \-\-stderr \|]
Close standard error
.TP
.RB [\| \-V | \-\-verbose \|]
Verbose message (IGNORED)
.TP
.B \| PROG \|
.RB [\| ARGS \|]
Program to execute with optional argument

.TP
.RB \| \-d | \-\-debug \|
.TP
.RB \| \-v | \-\-version \|
Print the version message
.TP
.RB \| \-h | \-\-help \|
Print the help message

.SH "RETURN VALUE"
.TP
.B 0
success
.TP
.B 1
miscellaneous operational failures

.SH "APPLICATION USAGE"
None.
.SH "FUTURE DIRECTIONS"
None.
.SH "SEE ALSO"
.BR svd (1),
.BR svl (1),
.BR svr (1),
.BR supervision (1),
.BR sv-init (8),
.BR sv-run (8),
.BR sv-rcorder (8),
.BR supervision (5),
.BR sv-shutdown (8)
.SH AUTHORS
tokiclover <tokiclover@supervision.project>
.\"
.\" vim:fenc=utf-8:ft=groff:ci:pi:sts=2:sw=2:ts=2:expandtab:
.\"


ChangeLog
Header: supervision/ChangeLog, 2018/08/16

0.14.7:

Memory alignement fix (fix structures padding);
Fix miultiple logfile open (important fix);
Added a workaround to LVM pvscan(8) stall with udev;

0.14.6:

Fix redundant dependency start up;
Fix logging to log file (important fix);
Fix zramfs infinite loop (0.14.5);

0.14.5:

Renamed loopback service to loop-back-device;

Tiny tixes to try to make LVM works out with udev.
Synchronization issue may happen because udev is not ready to initialize
devices; and then LVM will block the boot process.
Remain hitting the ^C (CTRL-C) to halt lvm.

0.14.4:

Fixed a nasty structure initilization bug (sv-run);

0.14.3:

Bug fixes: most important is svc_run() pid fix in sv-run(8);
Renamed `sv.log' logfile to `sv-rc.log';

0.14.2:

Merged remaining commits from master branch into this stable;
Some commits were missing from previous v0.14.1. **Apology**

0.14.1:

`sv-rc(8)' bug fix (environment variables were not being compared properly),
so SV_TIMEOUT was being defined with the value of SV_TIMEOUT_*, nasty bug;

0.14.0:

Removed `/lib/sv/opt' and merged the directory to `/lib/sv/sh';
Renamed `rs(8)' command to `sv-run(8)';
Renamed `sv-stage(8)' command to `sv-rc(8)'; along with the
environment variable `SV_STAGE' to `SV_INITLEVEL';
Renamed `OS_NAME' to `SV_UNAME' and added `SV_UNAME_RELEASE'
(defined in `${SV_TMPDIR}/env');
Renamed `runscript' shell script to `sv-run.sh';
Renamed `init-stage' shell script to `sv-init.sh';
Renamed `depgen' shell script to `sv-deps.sh';
Renamed `cmd' (supervision shell script) to `run';
Renamed `getty' service instances rightly with a `.' (dot)
instead of a hyphen;

*IMPORTANT NOTE**: if getty service are started, the sysadmin should add
the old symbolc links with
`for s in /etc/sv/getty-tty{2,3,4,5,6,S0,S1}; do ln -fs getty.tty1 $d; done'
before shutting down the service or system!


Manpages was cleaned and updated to v0.14.0;
See `supervision(5)' manual page for more information about the changes;
And also, `sv-run(8)' and `sv-rc(8)' manual pages;

A few bug fixes were ported to v0.13.1 (not all the bugs which require more
work; A nasty thread race bug was fixed lastly which was causing `sv-rc' to hang
waiting a SIGCHLD from a dead child;

`sv-config' was refactorized and coded in C; and now has a -u|--update command
line switch; a `sv-config --update' is now required when updating an old
install to handle the API/ABI changes from previous v0.13.0 and the changes
between v0.12.0 and v0.13.0;

The old `rs stage [COMMAND]' is now replaced with `sv-run init [COMMAND]'
which is now discouraged from general usage;
`sv-rc(8)' takes a runlevel argument like *standard* `rc(8)' utility;
The old `sv-rc --RUNLEVEL' (from previous release) is also supported;

New debugging code was added and can be enabled at runtime with `SV_DEBUG' or
`SVC_DEBUG' envronment variable;
An extra `SV_DEBUG_INFO' environement variable is available to service
information level message;
For more verbose debugging code, configure with `--enable-debug';
For logging debugging output: enable `SV_SYSLOG' and `SV_SYSLOG_INFO';
and then enable `SV_LOGGER' to save logging and debugging output to
`/var/log/sv.log';

ADDIDION:
SUPERVISION SERVICE:
RUNSCRIPT SERVICE: DEFAULT, SYSBOOT, local, qemu-vlan.vnet{0,4};

0.13.1:

Stabilisation release: there are fixes alll over the place, a few are
important and many are minors.

Important fixes:
- remove virtual service from dependency tree;
- remove failure and environment service state file before `start' command;
- `zap' command remove service pidfile;
- set up `__cmd__=finish' for supervision;
- added a (shell) `trap' for service state files;
- `fstabinfo' output any request;
- use standard GNU style `__attribute__';
- remove faulty `SV_PREFIX' from `sv-stage';
- and minor fixes, see `git log' for more;

0.13.0:

WARING: SHORT COMMAND LINE OPTION FOR RUNLEVELS CHANGED DRAMATICALLY FROM v.12.x

Improve signal handling so SIG{INT,QUIT,TERM} are not blocked anymore;
Improve parallelization of service (to avoid CHILD_MAX hard limit) and improve
scalabilty to thousends of service in parallel (POSIX threads);

Added support for keywords and timeout per service; so a NOHANG option is now
available to kill hangged service (see sv.conf for more iformation);
Keywords can be used to disable services per subsystem (like docker...);

Added support for containrization solutions or subsystems via keywords
usage (see KEYWORDS subsection in supervision(5) and sv.conf for more information)
for docker, LXC, jail, systemd-nspawn, prefix, supervision, UML, VServer and XEN.
Either the subsystem will be auto detected or use sv.conf to set a particular
subsystem with `SV_SYSTEM="${SUBSYSTEM}"' configuration variable...
`SV_SYSTEM="supervision"' for daemon supervision only;
`SV_SYSTEM="prefix"' for an isolated chrooted environment;
`SV_SYSTEM="docker"' for docker containers et al.
Services that have the subsystem keyword will not be started in that particular
subsystem environment.

To have the supervisor `({damontools[-encore],runit,s6})' executed as __PID 1__...
Just setup the container or subsystem; once done, use something like
the following for docker: `docker run [OPTIONS] --env container=docker --tmpfs /run
IMAGE /lib/sv/sh/init-stage --default'; and voila! the supervisor will be executed
as `PID 1' and another process will handle service management to setup the container.

Added abilty to abort an init-stage (used to abort and bring single runlevel
when checkfs service fails to clean up filesystems;

Added new manual to document new sv-stage(8) binary and refactorized rs(8);
Added a status command for init-stage/runlevel to retrieve service status:
e.g. `sv-stage --default status' to get service status of default runlevel;

Use POSIX threads to improve performance with parallel start/stop instead of
using only process based technics (child process/fork(2));


src/svp.c
/*
 * Copyright (c) 2016-2018 tokiclover <tokiclover@gmail.com>
 * This file is part of Supervision
 *
 * Supervision is free software; you can redistribute
 * it and/or modify it under the terms of the 2-clause, simplified,
 * new BSD License included in the distriution of this package.
 *
 * @(#)svp.c  0.15.0 2019/02/14
 */

#include <dirent.h>
#include <getopt.h>
#include <grp.h>
#include <pwd.h>
#include <sys/types.h>
#include <sys/resource.h>
#include <sys/file.h>
#include <fcntl.h>
#include "error.h"
#include "helper.h"
#include "sv-copyright.h"

#define SV_VERSION "0.15.0"

#define EXECVP(argv) do {                            \
	a = *argv;                                       \
	if ((p = strrchr(*argv, '/'))) *argv = p;        \
	execvp(a, argv);                                 \
	ERROR("Failed to execute `%s'", a);              \
} while (0/*CONSTCOND*/)

const char *progname;
char fb[BUFSIZ];
static struct slimit {
	unsigned long int sl_resource, sl_value;
	const char *sl_name, *__sl_ptr;
} SLIMIT[8] = {
	{ .sl_resource = RLIMIT_CORE, .sl_name = "core" },
	{ .sl_resource = RLIMIT_DATA, .sl_name = "data" },
	{ .sl_resource = RLIMIT_AS  , .sl_name = "memory" },
	{ .sl_resource = RLIMIT_NOFILE, .sl_name = "open" },
	{ .sl_resource = RLIMIT_FSIZE, .sl_name = "file" },
#ifdef RLIMIT_NPROC
	{ .sl_resource = RLIMIT_NPROC, .sl_name = "nproc" },
#endif
};
static char *group, *user, *pwnam;
static char *a, *p;
static int fd, o, rv;
static int f_lock = LOCK_EX;
static uid_t uid;
static gid_t gid;
static int f_egid, f_euid, f_env, id;
static struct passwd *pwd;
static struct group *grp;
static int f_exec;

static const char *shortopts = "012hPsVv/:b:c:d:e:f:G:g:L:l:m:n:o:p:U:u:";
static const struct option longopts[] = {
	{ "chroot" , 0, NULL, '/' },
	{ "name"   , 0, NULL, 'b' },
	{ "core"   , 0, NULL, 'c' },
	{ "data"   , 0, NULL, 'd' },
	{ "envdir" , 0, NULL, 'e' },
	{ "output" , 0, NULL, 'f' },
	{ "egid"   , 0, NULL, 'G' },
	{ "euid"   , 0, NULL, 'U' },
	{ "Lock"   , 0, NULL, 'L' },
	{ "lock"   , 0, NULL, 'l' },
	{ "memory" , 0, NULL, 'm' },
	{ "nice"   , 0, NULL, 'n' },
	{ "open"   , 0, NULL, 'o' },
	{ "proc"   , 0, NULL, 'p' },
	{ "gid"    , 0, NULL, 'g' },
	{ "uid"    , 0, NULL, 'u' },
	{ "sid"    , 0, NULL, 's' },
	{ "pgrp"   , 0, NULL, 'P' },
	{ "stdin"  , 0, NULL, '0' },
	{ "stdout" , 0, NULL, '1' },
	{ "stderr" , 0, NULL, '2' },
	{ "verbose", 0, NULL, 'V' },
	{ "help"   , 0, NULL, 'h' },
	{ "version", 0, NULL, 'v' },
	{ 0, 0, 0, 0 }
};
static const char *longopts_help[] = {
	"Set root directory",
	"Set first vector argument",
	"Set core size limit",
	"Set data size limit",
	"Use environment directory",
	"Set output size limit",
	"Set effective group identity",
	"Set effective user  identity",
	"Set exclusive lock file",
	"Set lock file",
	"Set memory limit",
	"Set process niceness",
	"Set open file limit",
	"Set process limit",
	"Set group identity",
	"Set user  identity",
	"Set session identity",
	"Set process group",
	"Close standard input",
	"Close standard output",
	"Close standard error",
	"Enable verbose output",
	"Print help message",
	"Print version string",
	NULL
};

static void envdir(int argc, char *argv[]);
static void setuidgid(int argc, char *argv[]);
static void setlock(int argc, char *argv[]);
static void softlimit(int argc, char *argv[]);
static void help_message(int retval);

__attribute__((__noreturn__)) static void help_message(int retval)
{
	printf("Usage: %s [OPTIONS] PROGRAM [ARGUMENTS]\n", progname);
	for ( o = 0; o < 16; o++)
		printf("    -%c, --%-6s  ARG    %s\n", longopts[o].val, longopts[o].name,
				longopts_help[o]);
	for ( ; longopts_help[o]; o++)
		printf("    -%c, --%-14s %s\n", longopts[o].val, longopts[o].name,
				longopts_help[o]);

	exit(retval);
}

static void envdir(int argc, char *argv[])
{
	DIR  *dp;
	int df, fd;
	struct dirent *de;

	if ((argc < 2)) exit(EXIT_FAILURE);
	if (!(dp = opendir(*argv)))
		ERROR("cannot open `%s' directory", *argv);
	df = dirfd(dp);

	while ((de = readdir(dp))) {
		if (*de->d_name == '.') continue;
		if ((fd = openat(df, de->d_name, O_RDONLY)) == -1) {
			ERR("cannot open `%s' file\n", de->d_name);
			continue;
		}
		
		rv = read(fd, fb, BUFSIZ);
		close(fd);
		if (rv == -1) continue;
		setenv(de->d_name, fb, 1);
	}

	argv++;
	if (!f_exec) return;
	EXECVP(argv);
}

static void setuidgid(int argc, char *argv[])
{
	if (f_exec) {
	static const char *shortopts = "G:g:U:u:";
	static const struct option longopts[] = {
		{ "egid"   , 0, NULL, 'G' },
		{ "euid"   , 0, NULL, 'U' },
		{ "gid"    , 0, NULL, 'g' },
		{ "uid"    , 0, NULL, 'u' },
		{ 0, 0, 0, 0 }
	};

	if (argc < 2) {
help_message:
		puts("Usage: %s [-U|-u USER:GROUP] [-G|-g GROUP] PROG [ARGS]");
		ERR("more arguments required!\n", NULL);
		exit(EXIT_FAILURE);
	}

	while ((o = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
		switch (o) {
		case 'G': f_egid++;
		case 'g': group = optarg; break;
		case 'U': f_euid++;
		case 'u': user = optarg; break;
		default: goto help_message;
		}
	}
	argc -= optind, argv += optind;
	}

	if (user) {
		pwnam = err_strdup(user);
		group = strchr(pwnam, ':');
		if (group)
			*group++ = '\0';
	}
	if (group)
		group = group;

	if (user) {
		if (sscanf(user, "%d", &id) == 1)
			pwd = getpwuid((uid_t) id);
		else
			pwd = getpwnam(user);
		if (pwd)
			uid = pwd->pw_uid;
		else {
			ERR("cannot get `%s' user identification\n", user);
			exit(EXIT_FAILURE);
		}
	}
	if (group) {
		if (sscanf(group, "%d", &id) == 1)
			grp = getgrgid((gid_t) id);
		else
			grp = getgrnam(group);
		if (grp)
			gid = grp->gr_gid;
		else {
			ERR("cannot get `%s' group identification\n", group);
			exit(EXIT_FAILURE);
		}
	}
	free(pwnam);

	if (uid) {
		if (setuid(uid)) ERROR("cannot set `%s' user  identification", user );
		if (f_env) {
			setenv("USER", user, 1);
			sprintf(fb, "%d", uid);
			setenv("UID", fb, 1);
		}
	}
	if (gid) {
		if (setgid(gid)) ERROR("cannot set `%s' group identification", group);
		if (f_env) {
			setenv("GROUP", group, 1);
			sprintf(fb, "%d", gid);
			setenv("GID", fb, 1);
		}
	}
	if (f_euid) {
		if (seteuid(uid))
			ERROR("cannot set `%s' effective user  identification", user );
	}
	if (f_egid) {
		if (setegid(gid))
			ERROR("cannot set `%s' effective group identification", group);
	}

	if (!f_exec) return;
	EXECVP(argv);
}

static void setlock(int argc, char *argv[])
{
	if (f_exec) {
	if (argc < 2) {
help_message:
		printf("Usage: %s [-n] FILE PROG [ARGS]\n", progname);
		exit(EXIT_FAILURE);
	}

	while ((o = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
		switch (o) {
		case 'n':
#ifdef HAVE_FLOCK
			f_lock |= LOCK_NB;
#else
			f_lock |= F_TLOCK;
#endif
			break;
		case 'N':
		case 'x':
		case 'X': break;
		default: goto help_message;
		}
	}
	argc -= optind, argv += optind;
	if (argc < 2) goto help_message;
	}
	
	if ((fd = open(*argv, O_CREAT | O_RDWR, 0644)) == -1)
		ERROR("cannot open `%s'", *argv);
	do {
#ifdef HAVE_FLOCK
		rv = flock(fd, f_lock);
#else
		rv = lockf(fd, f_lock, 0LU);
#endif
		if (rv == -1) {
			if (errno == EINTR) continue;
			ERROR("cannot lock `%s' file", *argv);
		}
	} while (rv);	
	argv++;

	if (!f_exec) return;
	EXECVP(argv);
}

static void softlimit(int argc, char *argv[])
{
	struct rlimit limit;

	if (f_exec) {
	if (argc < 2) {
help_message:
		printf("Usage: %s [-c ARG] [-d ARG] [-f ARG] [-m ARG] [-o ARG] [-p ARG] PROG [ARGS]\n", progname);
		exit(EXIT_FAILURE);
	}
	/* Parse options */
	while ((o = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
		errno = 0;
		switch (o) {
			case 'c':
				SLIMIT[0].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				break;
			case 'd':
				errno = 0;
				SLIMIT[1].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				break;
			case 'f':
				SLIMIT[4].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				break;
			case 'm':
				SLIMIT[2].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				break;
			case 'o':
				SLIMIT[3].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				SLIMIT[3].sl_value = SLIMIT[3].sl_value < _POSIX_OPEN_MAX ? _POSIX_OPEN_MAX : SLIMIT[3].sl_value;
				break;
			case 'p': /* IGNORED */
#ifdef RLIMIT_NPROC
				SLIMIT[5].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
#endif
				break;
			default: goto help_message;
		}
	}
	argc -= optind, argv += optind;
	if (argc < 1) goto help_message;
	}

	for (o = 0; SLIMIT[o].sl_name; o++) {
		if (SLIMIT[o].sl_value) {
			if (getrlimit(SLIMIT[o].sl_resource, &limit))
				ERROR("cannot get `%s' limit", SLIMIT[o].sl_name);
			limit.rlim_cur = limit.rlim_max < SLIMIT[o].sl_value ? limit.rlim_max :
				SLIMIT[o].sl_value;
			if (setrlimit(SLIMIT[o].sl_resource, &limit))
				ERR("cannot set `%s' limit\n", SLIMIT[o].sl_name);
		}
	}

	if (!f_exec) return;
	EXECVP(argv);

range_error:
	ERROR("invalid argument -- `%s'", argv[optind]);

	exit(EXIT_FAILURE);
}

#define ATOI(p) ((p[0] - '0')*10 + (p[1] - '0'))

int main(int argc, char *argv[])
{
	static int f_in, f_out, f_err, f_grp, f_sid;
	static int f_limit, f_nice, f_verbose;
	static char *env_dir, *work_dir;
	static char *f_arg, *lock;

	progname = strrchr(argv[0], '/');
	if (!progname)
		progname = argv[0];
	else
		progname++;

	if (argc < 1)
		help_message(1);

	if (strcmp(progname, "svp")) {
		argc--; argv++;
		f_exec++;
	}
	if (!strcmp(progname, "envdir")) {
		envdir(argc, argv);
	}
	else if (!strcmp(progname, "pgrphack")) {
		if (setpgid(0, 0) < 0) ERROR("cannot create a new session", NULL);
		EXECVP(argv);
		ERROR("Failed to execute `%s'", *argv);
	}
	else if (!strcmp(progname, "setsid")) {
		if (setsid() < 0) ERROR("cannot create a new session", NULL);
		EXECVP(argv);
		ERROR("Failed to execute `%s'", *argv);
	}
	else if (!strcmp(progname, "envuidgid")) {
		f_env++;
		setuidgid(argc, argv);
	}
	else if (!strcmp(progname, "setuidgid")) {
		setuidgid(argc, argv);
	}
	else if (!strcmp(progname, "setlock")) {
		setlock(argc, argv);
	}
	else if (!strcmp(progname, "softlimit")) {
		setuidgid(argc, argv);
	}

	/* Parse options */
	while ((o = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
		errno = 0;
		switch (o) {
			case '0': f_in++ ; break;
			case '1': f_out++; break;
			case '2': f_err++; break;
			case 'P': f_grp++; break;
			case 's': f_sid++; break;
			case 'V': f_verbose++; break;
			case '/':
				work_dir = optarg;
				break;
			case 'b':
				f_arg = optarg;
				break;
			case 'c':
				errno = 0;
				SLIMIT[0].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				f_limit++;
				break;
			case 'd':
				SLIMIT[1].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				f_limit++;
				break;
			case 'e':
				env_dir = optarg;
				break;
			case 'f':
				SLIMIT[4].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				break;
			case 'G': f_egid++;
			case 'g': group = optarg; break;
			case 'U': f_euid++;
			case 'u': user = optarg; break;
			case 'L':
#ifdef HAVE_FLOCK
				f_lock |= LOCK_NB;
#else
				f_lock |= F_TLOCK;
#endif
			case 'l':
				lock = optarg;
				break;
			case 'm':
				SLIMIT[2].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				f_limit++;
				break;
			case 'n':
				f_nice = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				break;
			case 'o':
				SLIMIT[3].sl_value = strtol(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
				f_limit++;
				SLIMIT[3].sl_value = SLIMIT[3].sl_value < _POSIX_OPEN_MAX ? _POSIX_OPEN_MAX : SLIMIT[3].sl_value;
				break;
			case 'p': /* IGNORED */
#ifdef RLIMIT_NPROC
				SLIMIT[5].sl_value = strtoul(optarg, (char**restrict)0, 10);
				if (errno == ERANGE) goto range_error;
#endif
				break;
			case 'v':
				printf("%s version %s\n\n", progname, SV_VERSION);
				puts(SV_COPYRIGHT);
				exit(EXIT_SUCCESS);
			case 'h':
				help_message(0);
				break;
			default:
				help_message(1);
				break;
		}
	}
	argc -= optind, argv += optind;
	if (argc < 1) {
		ERR("no enough arguments\n", NULL);
		exit(EXIT_FAILURE);
	}

	if (env_dir) {
		argc++; argv--;
		*argv = env_dir;
		envdir(argc, argv);
		argv++;
	}
	if (user || group) setuidgid(argc, argv); 
	if (lock) {
		argc++, argv--;
		*argv = lock;
		setlock(argc, argv);
		argv++;
	}
	if (f_limit) softlimit(argc, argv);
	if (f_nice) {
		errno = 0;
		if ((nice(f_nice) == -1) && errno) ERROR("cannot set `%d' niceness to process\n", f_nice);
	}
	if (work_dir) {
		if (chdir(work_dir))
			ERROR("cannot change current directory to `%s'", work_dir);
	}
	if (f_grp && (getpid() != getpgrp())) {
		if (setpgid(0, 0) < 0) ERROR("cannot create a new group", NULL);
	}
	if (f_sid) {
		if (setsid() < 0) ERROR("cannot create a new session", NULL);
	}
	if (f_in)  close(STDIN_FILENO);
	if (f_out) close(STDOUT_FILENO);
	if (f_err) close(STDERR_FILENO);

	EXECVP(argv);

range_error:
	ERROR("invalid argument -- `%s'", argv[optind]);

	return EXIT_FAILURE;
}
